### PR TITLE
Add an earlier status ('closed-out') to fix the parentage dependencies

### LINKED ISSUE #9656: Fix StepChain parentage for workflows in closed-out as well
**Impact of the new feature**
WMAgent & Unified

**Is your feature request related to a problem? Please describe.**
Given the asynchronous nature of merge execution in StepChain workflows, it could be that its output datasets get announced before all the parentage information has been updated.
The Parentage dependency is resolved through a ReqMgr2 cherrypy thread, which queries DBS, asking for every block in a dataset with no Parent block : [1]. Those are extremely heavy queries to the DBS database and sometimes due to external pressure to the system can lead to delays. When those delays exceed the 5 min timeout (configured in the cmsweb Frontend) the current query is dropped and the Parentage resolution of the current block is retried during the next cycle. This by its own means  can delay the process even further, while in the meantime Unified can see that  the needed statistics for a workflow is reached and announce the data. 
There is already a flag in the request description `ParentageResolved` which is used to sign StepChains with complete parentage information.
We need to fix parentage before it gets announced (closed-out status in ReqMgr2). Set that flag to `True`, and then propagated and take it into account by Unified. 

[1]
1. Given a dataset, find its parent datasets
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/DBS/DBS3Reader.py#L870

2. Find all the blocks in the dataset that do not have a parent block:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/DBS/DBS3Reader.py#L968

3. Given a block name, list files parents by lumi:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/DBS/DBS3Reader.py#L953

[2]
https://cmsweb-testbed.cern.ch/reqmgr2/fetch?rid=amaltaro_SC_LumiMask_PhEDEx_HG2004_Val_200330_093128_1213

**Describe the solution you'd like**
We should enable the `closed-out` status in the `StepChainParentageFixTask` cherrypy thread, such that Unified can check the `ParentageResolved` value before announcing StepChain workflows.
 
**Describe alternatives you've considered**
Other alternatives  are to  have WMAgent to only publish full information to DBS - this requiters long term development plans and redesign of portions of the WMAgent

**Additional context**
There is a related issue in Unified:
https://github.com/CMSCompOps/WmAgentScripts/issues/545

### PR DESCRIPTION
**Pull request by todor-ivanov**
Fixes #9656 

#### Status
ready

#### Description
Add an earlier status ('closed-out') in the list of statuses  to be considered when fixing the parentage dependencies

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
There should be one in the Unified repository. 

#### External dependencies / deployment changes
The flag `ParentageResolved`, which now (once this PR is merged) should be set in the `closoed-out` status, should be checked and taken into account by Unified.

### CODE REVIEW

**Comment by amaltaro** [comment]
Todor, can you please cherry-pick this PR as well:
https://github.com/dmwm/WMCore/pull/9538

note that the `insertFlag` has to be changed back to True though. Once those changes are here, we can close that issue.

File: src/python/WMCore/ReqMgr/CherryPyThreads/StepChainParentageFixTask.py [hunk 1]
```diff
@@ -65,8 +67,9 @@ def fixStepChainParentage(self, config):
         totalChildDS = len(childDatasets)
         fixCount = 0
         for childDS in childDatasets:
+            self.logger.info("Resolving parentage for dataset: %s", childDS)
             start = int(time.time())
-            failedBlocks = self.dbsSvc.fixMissingParentageDatasets(childDS, insertFlag=True)
+            failedBlocks = self.dbsSvc.fixMissingParentageDatasets(childDS, insertFlag=False)  # FIXME: set to True
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/CherryPyThreads/StepChainParentageFixTask.py:72]: set to True and remove the comment please ;)

**Comment by todor-ivanov** [comment]
@amaltaro , I just cherry-picked it. Not squashing those two commits though. Better have them separate.

**Review by amaltaro — CHANGES_REQUESTED** [review]
You didn't ask for it yet, but here goes my comment ;)

File: src/python/WMCore/ReqMgr/CherryPyThreads/StepChainParentageFixTask.py (hunk 1 above)
Comment (todor-ivanov) [reply src/python/WMCore/ReqMgr/CherryPyThreads/StepChainParentageFixTask.py]: Done :)

**Comment by amaltaro** [comment]
Thanks, Todor. It looks good to me.

**Review by amaltaro — APPROVED** [review]

**Pull request merged**